File: bundles/org.eclipse.vjet.vsf.vjo.kernel/src/org/eclipse/vjet/vsf/FlashTransport.java
/*******************************************************************************
 * Copyright (c) 2012 eBay Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     eBay Inc. - initial API and implementation
 *******************************************************************************/
package org.eclipse.vjet.vsf;

import org.eclipse.vjet.dsf.javatojs.anno.AJsProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsProxy;
import org.mozilla.mod.javascript.Scriptable;
import org.eclipse.vjet.dsf.javatojs.anno.AExclude;
import org.eclipse.vjet.dsf.dap.proxy.INativeJsFuncProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsFuncProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsTypeRef;
import org.eclipse.vjet.dsf.javatojs.anno.AJavaOnly;
import org.eclipse.vjet.vsf.ITransport;
import org.eclipse.vjet.vsf.Message;

//NativeJsProxy for vjo.dsf.FlashTransport.js
@org.eclipse.vjet.dsf.resource.utils.CodeGen("NativeJsProxyGenerator")
@AJsProxy
public class FlashTransport extends NativeJsProxy implements ITransport {

    /** for framework use only */
    @AExclude
    public FlashTransport(Scriptable nativeObj){
        super(nativeObj);
    }

    /** internal use only */
    protected FlashTransport(Object ...args){
        super(args);
    }

    public FlashTransport() {
        super();
    }

    public void handleRequest(Message message, Object callback) {
        callWithName("handleRequest", message, callback);
    }

    @AJavaOnly
    public static final NativeJsTypeRef<FlashTransport> prototype = NativeJsTypeRef.get(FlashTransport.class);

    public final INativeJsFuncProxy<FlashTransport> handleRequest = NativeJsFuncProxy.create(this, "handleRequest");
}

File: bundles/org.eclipse.vjet.vsf.vjo.kernel/src/org/eclipse/vjet/vsf/FlashTransport.js
/*******************************************************************************
 * Copyright (c) 2012 eBay Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     eBay Inc. - initial API and implementation
 *******************************************************************************/
vjo.ctype('org.eclipse.vjet.vsf.FlashTransport')
.satisfies('org.eclipse.vjet.vsf.ITransport')
.needs(['org.eclipse.vjet.vsf.Error','org.eclipse.vjet.vsf.Message','org.eclipse.vjet.vsf.SvcConfig','org.eclipse.vjet.vsf.SvcErr',
'org.eclipse.vjet.vjo.util.Map'])
.protos({
	//> public void handleRequest(Message message, Object callback)
	handleRequest : function(message, callback) {
		
		var swf = document.getElementById("flash_proxy");
		// if flash object is specified on the page, use		
		if (swf == null) {
			// Log error
			return;
		}
		var cfg = message.svcConfig; //< org.eclipse.vjet.vsf.SvcConfig
		var requestUrl = cfg.url; //< String
	   try {
			var contentType;
			if (cfg.reqtMarshalling == 'SOAP') {
				contentType = "text/xml";
			} else {
				contentType = "application/json";
			}
			var svcConfigObj = {};
			svcConfigObj['async'] = cfg['async'];
			svcConfigObj['method'] = cfg['method'];
			svcConfigObj['objType'] = cfg['objType'];
			svcConfigObj['reqtMarshalling'] = cfg['reqtMarshalling'];
			svcConfigObj['respMarshalling'] = cfg['respMarshalling'];
			svcConfigObj['timeout'] = cfg['timeout'];
			svcConfigObj['url'] = cfg['url'];
			// heck
			svcConfigObj['flash'] = true;
			svcConfigObj['typeMappingClassName'] = message.typeMappingClassName;
			svcConfigObj['serviceName'] = message.svcName;
			svcConfigObj['operationName'] = message.opName;

			var header = {};
			if (message.svcConfig.svcType == 'SOA') {
		  		 // gyue: add SOA specific headers
		   		header = new Array();
		   		var i = 0;
		   		// populate the header array from message.headers
		   		for (var m in message.headers) {
					header[i] = {"name":m, "value":message.headers[m]};
					i++;
		   		}
			}

			// create a swf complaint request
			var request = {  serviceUrl: requestUrl,
				headers: header,
				data: message.rawRequest,
				contentType: contentType,
				callbackSvcId: message.svcId};
			swf.callback = function(response){
				message.response = {};
				message.response.data = {};
				message.response.data.documentElement = response.firstChild;
				callback.apply(this, new Array(message));
			}
			swf.onRemoteCall({request: request, svcConfig: svcConfigObj});
	   } catch(e) {
			var error = new org.eclipse.vjet.vsf.Error(); //< org.eclipse.vjet.vsf.Error
			error.id = org.eclipse.vjet.vsf.SvcErr.InvRq;
			error.message="Cannot open URL '"+requestUrl+"'";
			message.response.addError(error);
	 	}
	},
	// TEMP: remove after swf callback fix
	receive : function(message) {
	
		var vjoMessage = new org.eclipse.vjet.vsf.Message(message.svcConfig.serviceName, message.svcConfig.operationName); 
		vjoMessage.typeMappingClassName = message.svcConfig.typeMappingClassName;
		
		// copy vjo characteristics 
		var result = {}; 
		for (var i in vjoMessage) { 
			message[i] = vjoMessage[i]; 
		}	
	}
})
.endType();